[develop] Common infrastructure for CIs and manual runs

danielabdi-noaa opened the pull request:
## DESCRIPTION OF CHANGES: 
This PR addresses the issue described in detail in issue #377 . Basically, it makes both CIs ( Jenkins and Github actions ) use the same build and test infrastructure, especially simplifying Jenkins a lot.

Fundamental test cases ( 9 cases used in Github actions CI) are all green[1] on all platform. This includes test besides build. Here is the [jenkins pipeline](https://jenkins-epic.woc.noaa.gov/blue/organizations/jenkins/ufs-srweather-app%2Fpipeline/detail/PR-378/8/pipeline/189) for future reference. I think this PR is high priority because of this.

[1] Although Cheyenne tests are red, all tests completed successfully on the system. There is some weird issue with tarring the results that does not happen on other systems.

**Update:** Cheyenne issues are fixed and here is the [latest jenkins pipeline](https://jenkins-epic.woc.noaa.gov/blue/organizations/jenkins/ufs-srweather-app%2Fpipeline/detail/PR-378/27/pipeline/158) result that succeeded on all platforms including Cheyenne.

Detailed description changes:

- Build for Jenkins is done with `test/build.sh`
- WE2E tests for Jenkins are run with `tests/WE2E/setup_WE2E_tests.sh`
- Configuration for both Jenkins and Github actions are done with files: `fundamental`, `comprehensive` or any other name such as `custom`. To create a configuration for a specific machine, for example in case of a test case that can not be successfully run on that machine, create config files of the form `fundamental.jet` etc
- Job status monitoring in Jenkins is done with existing script `tests/WE2E/get_expts_status.sh`

With these changes, Jenkins source code is simplified and more robust, because code is not duplicated and uses existing
mature scripts for build and test.

Moreover, bug fixes for Jenkins tests include:
- Bug fix for Gaea (pre-generated grid directory was not being set in `machine/gaea.sh`).
- Bug fix for Cheyenne. 
     - Conda was not being loaded because python was loaded in `build_cheyenne_intel/gnu` and then `wflow_cheyenne` did not unload python.
     - `set +u` needed for SLURM_JOB_ID check
- Orion issue seem to be suddenly gone without any change (maybe software update on the system fixed the issue)
- NoaaCloud still have issues with `libpng` version not found (disabled for now, will come back after issues are resolved)

I have left the selection of comprehensive test cases to get a green on all machines for a future PR. Jet is especially difficult since a test case may succeed or fail randomly on it for unknown reasons.
### Type of change
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
- [X] hera.intel
- [X] orion.intel
- [X] cheyenne.intel
- [X] cheyenne.gnu
- [X] gaea.intel
- [X] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [X] Jenkins
- [X] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
None

## DOCUMENTATION:
None

## ISSUE: 
#377 

## CHECKLIST
- [X] My code follows the style guidelines in the Contributor's Guide
- [X] I have performed a self-review of my own code using the Code Reviewer's Guide
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes do not require updates to the documentation (explain).
- [X] My changes generate no new warnings
- [X] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 
A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [X] enhancement
- [ ] documentation
- [ ] release
- [X] high priority
- [ ] run_ci
- [X] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

## CONTRIBUTORS (optional): 
@jessemcfarland @christinaholtNOAA

danielabdi-noaa committed:
Combine CI infrastructure, use get_expts_status etc to simplify Jenkins.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1066325011/20220925160516/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 1 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: custom_GFDLgrid
All experiments completed

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1066325011/20220925160511/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 1 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: custom_GFDLgrid

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1066325011/20220925172012/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 1 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1066325011/20220925172009/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 1 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE

danielabdi-noaa commented:
REBUILD

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1066325011/20220925182009/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta
All experiments completed

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/1066325011/20220925182013/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
Experiment Succeeded on jet: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on jet: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
Experiment Succeeded on jet: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta

danielabdi-noaa commented:
REBUILD

danielabdi-noaa commented:
REBUILD

danielabdi-noaa commented:
REBUILD

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1066325011/20220926012010/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 1 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE

danielabdi-noaa commented:
REBUILD

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1066325011/20220926025012/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: nco_grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta
All experiments completed

danielabdi-noaa commented:
REBUILD

MichaelLueken commented:
@danielabdi-noaa @jessemcfarland - I'm just wanting to check whether this PR supersedes PR #333.

danielabdi-noaa commented:
@MichaelLueken I did start off by importing all changes from PR #333 to try and fix the issues described in the discussion there, but ended up doing more to try and make both CIs use same infractructure. Not sure if it supersedes the PR since it looks like there is the "tarring log files issue" on Cheyenne that maybe @jessemcfarland can help fix?

danielabdi-noaa committed:
Unload python in wflow_cheyenne just to be safe.

danielabdi-noaa committed:
Don't do launch_wflow when checking experiment status.

danielabdi-noaa commented:
I've rebased the PR on develop now. Please let me know if you see any issues or have suggestions on how to avoid potential cron issues: maybe increase cron interval to 4 min,  increase the poll frequency from 1min to 2-3min etc?

Commit:
Debug post test tar failures on Cheyenne

Commit:
Disable PW AWS cluster for pipeline

danielabdi-noaa committed:
Fake un-launched jobs as if they are in progress.

Commit:
Fix variable expansion for the srw_test.sh call

JeffBeck-NOAA commented:
> I've rebased the PR on develop now. Please let me know if you see any issues or have suggestions on how to avoid potential cron issues: maybe increase cron interval to 4 min, increase the poll frequency from 1min to 2-3min etc?

@danielabdi-noaa, yes, I think we should at least set the cron interval to 4-5 minutes.

danielabdi-noaa committed:
Increase cron interval to 5 mins.

danielabdi-noaa committed:
Undo last commit.

danielabdi-noaa committed:
Also increase initial delay to 6 min.

venitahagerty commented:
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1066325011/20221001145011/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE

Commit:
Updates for build artifact handling

* Update tar command to create cleaner artifact
* Update path of build log for s3 upload

Commit:
Remove lines used for debugging

danielabdi-noaa commented:
The issue on Cheyenne has be resolved now as the last Jenkins test shows. So all platforms with the exception of noaacloud, which is temporarily disabled, pass the fundamental tests with green. I think it is ready to be merged.

MichaelLueken commented on tests/WE2E/machine_suites/custom:
```diff
@@ -0,0 +1 @@
+custom_GFDLgrid
```
@danielabdi-noaa Should we keep this custom tests file, or would it be best to remove it before merging the PR?

danielabdi-noaa committed:
Remove custom test list.

danielabdi-noaa committed:
Add custom_ensemble_2mems_stoch.

danielabdi-noaa committed:
Undo unloading of python for Cheyenne.

MichaelLueken commented:
@danielabdi-noaa Thank you very much for removing the `machine_suites/custom` file from the batch of changes to be merged!

I see from your last push, that you removed the unloading of python for Cheyenne in `wflow_cheyenne`.  From the description of changes, specifically bugs that were identified and corrected, I saw the following for Cheyenne:

- Conda was not being loaded because python was loaded in build_cheyenne_intel/gnu and then wflow_cheyenne did not unload python.

Should the unloading of python from `wflow_cheyenne` be kept due to the above bug?

danielabdi-noaa commented:
@MichaelLueken Cheyenne is kind of complicated because at the moment it can not load conda if the user has python loaded.
For example, the following will not work on Cheyenne because the build modulefiles load python, while `wflow_cheyenne` does not unload python.
```
module use modulefiles
module load build_cheyenne_intel
module load wflow_cheyenne
```
PR #333 was doing this which was why it was not able to get to build stage. I fixed the script to skip loading the build modulefile, so we may or maynot have the `module unload python` in `wflow_cheyenne`. It was a safety precaution to add it there. However, there is a problem with unloading python, because doing so leaves you with system python2 (not python3) which causes cron tab issues. We think that is the case, but we will test it out in the afternoon to make sure that is the case.

Commit:
debug cheyenne python failures

danielabdi-noaa committed:
Add more debugging code for cheyenne.

danielabdi-noaa committed:
Decrease re-launch interval temporarily.

danielabdi-noaa committed:
Turn off check_var_valid.

danielabdi-noaa committed:
Add Lmod_init for Gaea.

danielabdi-noaa committed:
Clean up now that everything seems to work.

MichaelLueken commented on ush/launch_FV3LAM_wflow.sh:
```diff
@@ -151,6 +151,19 @@ module load "${WFLOW_MOD_FN}" > /dev/null 2>&1 || print_err_msg_exit "\
+#
+#-----------------------------------------------------------------------
+#
+# Hack for Gaea since system python version is 3.4.10
```
Is this a hack for Gaea or a hack for Cheyenne?  I just want to check because the logic is for Cheyenne, not Gaea (the directory structure is also Cheyenne).

danielabdi-noaa replied:
@MichaelLueken You are correct, it should read for Cheyenne!
I think best would be to update the system python version on Cheyenne to atleast v3.6. Will open an issue for that later. The SRW app requirement is not clear because `environment.yaml` specifies python version of atleast 3.7 and in the code, there is a requirement of atleast 3.6 in `generate_workflow.py`. These seem to be requirements for the python version in coda specifically, not for the system python. The transition from shell to python workflow generation has imposed system python requirement of atleast version 3.6, so this need to be spelled out in the document. All platforms had system python version of >3.6 except Cheyenne that had `3.4.10`. This was the cause for the crontab entry deletion issue.

danielabdi-noaa replied:
I've opened up the issue here #402

MichaelLueken replied:
@danielabdi-noaa Thanks!

MichaelLueken reviewed (commented):
@danielabdi-noaa One quick question with the modification in `ush/launch_FV3LAM_wflow.sh` - the comment states that this change is a hack for Gaea, but the logic is for Cheyenne, with a Cheyenne directory.  Should the comment read:

`Hack for Cheyenne`

instead?

danielabdi-noaa committed:
Fix comment about the need for conda activation for Cheyenne.

danielabdi-noaa commented:
@MichaelLueken I think this PR is now ready to go in. All the fundamental tests completed successfully on Cheyenne with both gnu and intel. All other systems (except Hera) have green too on fundamental tests. Hera is unusually slow today and I've just checked the jobs are queued. It will probably finish with green by tomorrow. I will push the commit fixing some comments so here is the Jenkins pipeline for future reference. I've put it under the description of the PR as well

[Jenkins fundamental test run](https://jenkins-epic.woc.noaa.gov/blue/organizations/jenkins/ufs-srweather-app%2Fpipeline/detail/PR-378/27/pipeline/158)

**Edit:** All machines have finished successfully with fundamental tests now.

danielabdi-noaa committed:
Print crontab content before/after deletion.

MichaelLueken reviewed (commented):
Thank you very much for working through the issues on the tier-1 machines, @danielabdi-noaa and @jessemcfarland!  I have submitted the fundamental WE2E tests on Hera and they all passed successfully.  Before submitting my approval, I see this morning's Jenkins run has failed - https://jenkins-epic.woc.noaa.gov/blue/organizations/jenkins/ufs-srweather-app%2Fpipeline/detail/PR-378/28/pipeline.  Will this need to be addressed before approval and merging, or is this an issue with either GitHub or Jenkins?

danielabdi-noaa commented:
@MichaelLueken Yes, it was a github issue. There is an ongoing re-run that should finish in an hour or so.

Commit:
Improve experiments directory clean up

Only remove the data directories to allow we2e cron jobs to complete
and clean up themselves correctly.

Commit:
Ensure we2e cron jobs execute again before cleanup

christinaholtNOAA commented on test/build.sh:
```diff
@@ -99,63 +90,63 @@ declare -a executables_created=( chgres_cube \
+    # Array of all optional rrfs_utl executables built
+    #-----------------------------------------------------------------------
+    executables_created=( adjust_soiltq.exe \
+                          check_imssnow_fv3lam.exe \
```
Should this one also be a `+=`?

danielabdi-noaa replied:
Thanks, just fixed the bug!

christinaholtNOAA commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -1040,7 +1040,7 @@ def generate_FV3LAM_wflow():
             edit the cron table):
 
-            */3 * * * * cd {EXPTDIR} && ./launch_FV3LAM_wflow.sh called_from_cron=\"TRUE\"
+            */{CRON_RELAUNCH_INTVL_MNTS} * * * * cd {EXPTDIR} && ./launch_FV3LAM_wflow.sh called_from_cron=\"TRUE\"
```
NICE!

christinaholtNOAA reviewed (approved):
This is a really nice addition, @danielabdi-noaa. Thanks!

Just one possible correction below.

danielabdi-noaa committed:
Bug fix in build.sh.

MichaelLueken reviewed (approved):
The rerun has of the Jenkins pipeline has successfully completed.  I will now give my approval to these changes.
